### PR TITLE
DKG Resharing

### PR DESCRIPTION
**Pull request by gakonst**
- Cherrypicks the feature from https://github.com/celo-org/celo-threshold-bls-rs/pull/74 (I still need to review the paper and the code in-depth)
- Removes the `node.rs::PhaseX` structs and just impls the `DKGPhase` trait for any of the individual `PhaseX` implementers. This allows using the node with both the JF and the Resharing mode and will only require modifications on the CLI side (future PR)

### CODE REVIEW

File: crates/dkg-cli/src/user/actions.rs [hunk 1]
```diff
@@ -1,20 +1,22 @@
 use super::opts::{FinalizeOpts, NewOpts, PublishSharesOpts, StateOpts};
-use crate::CLIResult;
 use rand::RngCore;
 use std::{fs::File, io::Write};
 
 use dkg_core::{
-    node::{DKGPhase, Phase0, Phase1, Phase2, Phase2Result, Phase3},
+    dkg_impls::joint_feldman::{DKGWaitingJustification, DKGWaitingResponse, DKGWaitingShare, DKG},
```
Comment (nikkolasg) [inline crates/dkg-cli/src/user/actions.rs:6]: I have to say I dont like that at all - yet another import. DKG lib is already separated into two folders and now there is a third one... We're very far from simplicity now.

File: crates/dkg-core/src/node.rs [hunk 2]
```diff
@@ -46,103 +43,58 @@ pub trait DKGPhase<C: Curve, B: BoardPublisher<C>, T> {
...
-impl<C, B, R> DKGPhase<C, B, &mut R> for Phase0<C>
+impl<C, B, R, P> DKGPhase<C, B, &mut R> for P
 where
     C: Curve,
     B: BoardPublisher<C>,
     R: RngCore,
+    P: Phase0<C>,
 {
-    type Next = Phase1<C>;
+    type Next = P::Next;
```
Comment (nikkolasg) [inline crates/dkg-core/src/node.rs:53]: <3 I love that integration !

File: crates/dkg-cli/src/user/actions.rs (hunk 1 above)
Comment (gakonst) [reply crates/dkg-cli/src/user/actions.rs]: wdyt about https://github.com/celo-org/celo-threshold-bls-rs/pull/79/commits/146735bfb3fc4b830842c0f1a7b65ff36a398ac7?

File: crates/dkg-core/src/node.rs [hunk 3]
```diff
@@ -1,16 +1,13 @@
 use super::{
     board::BoardPublisher,
-    dkg_impls::joint_feldman::{DKGWaitingJustification, DKGWaitingResponse, DKGWaitingShare, DKG},
     primitives::{
-        group::Group,
-        phases::{Phase0 as DPhase0, Phase1 as DPhase1, Phase2 as DPhase2, Phase3 as DPhase3},
+        phases::{Phase0, Phase1, Phase2, Phase3},
```
Comment (nikkolasg) [inline crates/dkg-core/src/node.rs:4]: I love it <3

**Pull request merged**